---
title: "DGE analysis of TCGA data"
output: html_notebook
---


Set up the necessary libraries & downloads
```{r}
library(TCGAbiolinks) 
library(SummarizedExperiment)
library(BiocManager)
library(readxl)
library(airway)
library(DESeq2)
library(pheatmap) 
library(RColorBrewer)
library(readr)
library(dplyr)
library(plyr)
setwd("~/Google Drive/ISSI Project/Noah's Code")
```

Make a summarised-experiment using a query
```{r}
query <- GDCquery(
  project = 'TCGA-ESCA', # This is oesophageal cancer
  data.category = 'Transcriptome Profiling',
  data.type = 'Gene Expression Quantification',
  experimental.strategy = 'RNA-Seq',
  workflow.type = 'HTSeq - Counts',
  legacy = FALSE
)
GDCdownload(query)
se <- GDCprepare(query, summarizedExperiment = TRUE)

```

Explore the available information
```{r}
dim(rowData(se)) #56493 transcripts, 3 pieces of information. The rows are the genes, the columns are the information
#rownames(se) #the names of each transcript
names(rowData(se)) #data for each row: a row corresponds to a gene transcript, of which there are around 56,000. Each gene has ensembl_gene_id, external_gene_name, original_ensembl_gene_id


dim(colData(se)) #173 samples, 141 pieces of information for each. The column titles are the rows, and the data about each column titile are the columns of this data frame
#colnames(se) #the barcode of each patient
#colData(se) #data for each column: a column corresponds to a cell sample. Each sample has a lot of data about the patient, etc.
names(colData(se)) #what info is available

#head(assay(se)) #contains the actual assay info (row=gene transcript, column=patient)


table(table(colData(se)$patient)==2)
```

```{r}
table(colnames(assay(se)) == colData(se)$barcode) #the names of the assay columns (the patient barcodes) are all in the same order as the column data, helpfully

table(rownames(assay(se)) == rowData(se)$ensembl_gene_id) #true for all of these as well
```
get the patients we care about
```{r}
#get the sample type code using the barcode
colData(se)$Sample_Type_Code <- substr(se$barcode, 14, 15) # Sample Type Code is found in the 14th and 15th characters of the barcode


#create names for these
colData(se)$Sample_Type <- sapply(se$Sample_Type_Code,
                         function(x) ifelse(x == '01','PrimaryTumour', ifelse(x %in% c('10','11'),'Normal','Other')))

#select only the sample types we care about (PrimaryTumor and Normal)
se <- se[,se$Sample_Type %in% c('PrimaryTumour','Normal')] 

#check it worked
table(colData(se)$Sample_Type) #it did
```


```{r}
#I've re-ordered the levels on the Sample_Type so that "Normal" is the base level. The second line is actually redundant but left it in in case someone wants to chage the order
se$Sample_Type <- factor(se$Sample_Type, levels = c("Normal", "PrimaryTumour"))
se$Sample_Type <- relevel(se$Sample_Type, "Normal")
table(colData(se)$Sample_Type)
```
```{r}
#creating a list of patients with 2 samples 


se_onlydoubles <- se[, duplicated(colData(se)$patient)]

dim(colData(se_onlydoubles))

table(se_onlydoubles$Sample_Type)
```

some non-normalised PCA
```{r}
dds <- DESeqDataSet(se, design = ~ patient + Sample_Type)
trial <- DESeqTransform(dds)


dds_paired <- DESeqDataSet(se_onlydoubles, design = ~ Sample_Type)
trial_paired <- DESeqTransform(dds_paired)

```

normalisation (varianceStabilizingTransform() has too large a runtime)
```{r}
dds <- estimateSizeFactors(dds)
counts_normalized <- counts(dds, normalized = TRUE)
counts_log_normalized <- log2(counts_normalized + 1)


dds_paired <- estimateSizeFactors(dds_paired)
counts_normalized_paired <- counts(dds_paired, normalized = TRUE)
counts_log_normalized_paired <- log2(counts_normalized_paired + 1)
```


```{r}
pdf(file="figs/PCA_unpaired.pdf")
dds_normalised_log <- dds
assay(dds_normalised_log) <- counts_log_normalized
trial_normalised_log <- DESeqTransform(dds_normalised_log)
plotPCA(trial_normalised_log, intgroup = 'Sample_Type')
dev.off()

pdf(file="figs/PCA_paired.pdf")
dds_paired_log<-dds_paired
assay(dds_paired_log) <- counts_log_normalized_paired
trial_normalised_log_paired <- DESeqTransform(dds_paired_log)
plotPCA(trial_normalised_log_paired, intgroup = 'Sample_Type')
dev.off()
```

Loading up genes of interest, and run the same basic normalisation & analysis on them
```{r}
e2f_targets <- read.table('E2F.GeneTargets.txt', header=TRUE) #imports a list of target genes
se_E2F <- subset(se, rowData(se)$external_gene_name %in% e2f_targets$external_gene_name)


colData(se_E2F)$Sample_Type_Code <- substr(se$barcode, 14, 15)
colData(se_E2F)$Sample_Type <- sapply(se_E2F$Sample_Type_Code,
                         function(x) ifelse(x == '01','PrimaryTumour', ifelse(x %in% c('10','11'),'Normal','Other')))
se_E2F <- se_E2F[,se_E2F$Sample_Type %in% c('PrimaryTumour','Normal')] 

table(colData(se_E2F)$Sample_Type) #it did

#I've re-ordered the levels on the Sample_Type so that "Normal" is the base level. The second line is actually redundant but left it in in case someone wants to chage the order
se_E2F$Sample_Type <- factor(se_E2F$Sample_Type, levels = c("Normal", "PrimaryTumour"))
se_E2F$Sample_Type <- relevel(se_E2F$Sample_Type, "Normal")
table(colData(se_E2F)$Sample_Type)



se_E2F_paired <- se_E2F[, duplicated(colData(se_E2F)$patient)]
dds_E2F_paired <- DESeqDataSet(se_E2F, design = ~ patient + Sample_Type)

```


Running DGE on E2F genes
```{r}

se_E2F_paired <- se_E2F[, duplicated(colData(se_E2F)$patient)]
dds_E2F_paired <- DESeqDataSet(se_E2F_paired, design = ~ Sample_Type)
dds_E2F_DGE_paired <- DESeq(dds_E2F_paired)
```

```{r}
dds_E2F <- DESeqDataSet(se_E2F, design = ~ patient + Sample_Type)


dds_E2F_DGE<- DESeq(dds_E2F)
```
normal results
```{r}
dds_E2F_DGE_results <- results(dds_E2F_DGE) # This extracts the results from the DESeqDataSet object dds_DGE
head(dds_E2F_DGE_results)

write.table(dds_E2F_DGE_results, file = 'DESeq2_Sig_results_cancer_Vs_normal.txt',
            sep = '\t', quote = FALSE, row.names = TRUE) # This writes the table so you can load it somewhere else
```
histogram
```{r}
pdf(file="figs/histogram.pdf")
hist(dds_E2F_DGE_results$padj, xlab = 'adjusted p-value', main = 'frequency of p-values', nclass=100)
dev.off()
```
```{r}
pdf(file="figs/volcano.pdf")
plot(x = dds_E2F_DGE_results$log2FoldChange,
     y = -log10(dds_E2F_DGE_results$padj),
     pch = 16, cex = 0.5,
     xlab = 'log2(Fold Change)', ylab = '-log10(adjusted p-value)',
     main = 'Volcano plot',
     col=ifelse((dds_E2F_DGE_results$log2FoldChange>2 | dds_E2F_DGE_results$log2FoldChange< -2)&dds_E2F_DGE_results$padj<0.01 , "red", "black"))
dev.off()
```


```{r}
dds_E2F_DGE_paired_results <- results(dds_E2F_DGE_paired) # This extracts the results from the DESeqDataSet object dds_DGE
head(dds_E2F_DGE_paired_results)

write.table(dds_E2F_DGE_paired_results, file = 'DESeq2_Sig_results_cancer_Vs_normal_paired.txt',
            sep = '\t', quote = FALSE, row.names = TRUE) # This writes the table so you can load it somewhere else
```


```{r}
#BiocManager::install('biomaRt')
library(biomaRt)

ensembl = useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl")
```


```{r}
genes_to_search=rownames(dds_E2F_DGE_results)

searched_genes <- getBM(attributes=c('ensembl_gene_id','hgnc_symbol','chromosome_name','start_position','end_position'),filters = 'ensembl_gene_id', values = genes_to_search, mart = ensembl)

searched_genes <- distinct(searched_genes, ensembl_gene_id, .keep_all = TRUE)

#dim(searched_genes)

dds_E2F_DGE_results$chromosome_name<-searched_genes$chromosome_name
dds_E2F_DGE_results$start_pos<-searched_genes$start_position
dds_E2F_DGE_results$end_pos<-searched_genes$end_position
dds_E2F_DGE_results$gene_symbol<-searched_genes$hgnc_symbol

dds_E2F_DGE_results$chromosome_number[dds_E2F_DGE_results$chromosome_name=="X"]<-"23"

dds_E2F_DGE_results$chromosome_number = as.numeric(dds_E2F_DGE_results$chromosome_number)

dds_E2F_DGE_results <- subset(dds_E2F_DGE_results, select= -c(chromosome_name, pvalue))

dds_E2F_DGE_results$chromosome_length <- mapvalues(dds_E2F_DGE_results$chromosome_number, from=c(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23), to =c(248956422, 242193529, 198295559, 190214555, 181538259, 170805979, 159345973, 145138636, 138394717, 133797422, 135086622, 133275309, 114364328, 107043718, 101991189, 90338345, 83257441, 80373285, 58617616, 64444167, 46709983, 50818468, 156040895))

dds_E2F_DGE_results$chromosomal_position <- dds_E2F_DGE_results$chromosome_number + dds_E2F_DGE_results$start_pos/dds_E2F_DGE_results$chromosome_length
```

Manhattan plot
```{r}
#dds_E2F_DGE_results
#pdf(file="figs/manhattan.pdf")
plot(x = dds_E2F_DGE_results$chromosomal_position-0.5,
     y = -log10(dds_E2F_DGE_results$padj),
     pch = 16, cex = 0.5,
     xlab = 'chromosomal position', ylab = '-log10(adjusted p-value)',
     main = 'Manhattan plot',
     col = ifelse(dds_E2F_DGE_results$chromosome_number %in% c(1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23), "dark grey", "black")
     )
axis(1, (1:23))


```


Significant genes histogram by chromosome
```{r}
significant_genes = na.omit(dds_E2F_DGE_results)

significant_genes = significant_genes[significant_genes$padj<0.01, ]
significant_genes = significant_genes[(significant_genes$log2FoldChange<-2 | significant_genes$log2FoldChange>2), ]



```


```{r}
significant_genes
write.table(dds_E2F_DGE_results, file = 'DESeq2_Sig_results_cancer_Vs_normal.csv',
            sep = ',', quote = FALSE, row.names = TRUE)
write.table(significant_genes, file = 'significant_genes.csv',
            sep = ',', quote = FALSE, row.names = TRUE)
#hist(dds_E2F_DGE_results$chromosomal_position, nclass=23, main="Histogram of Significant Genes")
#axis(1, (1:23))
```

